File: kotlin/src/_20_valid_parentheses.kt
/*
 * @lc app=leetcode id=20 lang=kotlin
 *
 * [20] Valid Parentheses
 */

// @lc code=start
class Solution {
    fun isValid(s: String): Boolean {
        val stack = ArrayDeque<Char>()
        val pairs = mapOf(')' to '(', ']' to '[', '}' to '{')
        for (c in s) {
            val open = pairs[c]
            if (open != null) {
                if (stack.isEmpty() || stack.removeLast() != open) return false
            } else {
                stack.addLast(c)
            }
        }
        return stack.isEmpty()
    }
}
// @lc code=end

File: typescript/src/_20_valid_parentheses.ts
/*
 * @lc app=leetcode id=20 lang=typescript
 *
 * [20] Valid Parentheses
 */

// @lc code=start
function isValid(s: string): boolean {
    const stack: string[] = [];
    const pairs = new Map([[')', '('], [']', '['], ['}', '{']]);
    for (const c of s) {
        const open = pairs.get(c);
        if (open !== undefined) {
            if (stack.pop() !== open) return false;
        } else {
            stack.push(c);
        }
    }
    return stack.length === 0;
};
// @lc code=end
